ornlneutronimaging/mcpevent2hist
# Add new fast sophiread library for high-throughput data processing

# Description of Pull Request

This PR introduces a new sub-lib, `FastSophireadLib` that can support high throughput data processing via multiple threading and local optimization.

## Purpose of work

The purpose of this work is to provide a new set of class and functions that can sustain high data throughput during experiment.



## Summary of work

- New class `Hit` with special constructor for in-place construction
- New struct `TPX3` for holding data from one chip per burst
- New set of functions for processing data in high speed.

## Additional detail of work


## Testing instructions

- update the environment to include both `spdlog` and `tbb`
- configure the project with cmake as usual
- make the project following existing instructions
- run `./FastSophireadBenchmarks_raw2hits.app data/frames_flood_1M.tpx3` to perform benchmark on target system

```bash
❯ ./FastSophireadBenchmarks_raw2hits.app data/frames_flood_1M.tpx3 
[2023-09-01 16:00:56.247] [info] File size: 77916776 bytes
[2023-09-01 16:00:56.275] [info] Read raw data: 0.029082 s
[2023-09-01 16:00:56.275] [info] ***Single thread processing***
[2023-09-01 16:00:56.464] [info] Number of hits: 9647818
[2023-09-01 16:00:56.464] [info] Single thread processing: 0.157781 s
[2023-09-01 16:00:56.464] [info] Single thread processing speed: 6.114689e+07 hits/s
[2023-09-01 16:00:56.464] [info] ***Multi-thread processing***
[2023-09-01 16:00:56.605] [info] Number of hits: 9647818
[2023-09-01 16:00:56.605] [info] Multi-thread processing: 0.10753499999999999 s
[2023-09-01 16:00:56.605] [info] Multi-thread processing speed: 8.971793e+07 hits/s
```
> note: the actual number is system dependent.
- run `ctest` for verify new unit tests.
